Cgen doesn't unwrap sumtype to another

### Describe the bug

<title>

### Reproduction Steps

```v
import x.json2

type Foo = string | int | f64

fn f(j map[string]json2.Any) {
	assert (j['name'] or { return } as string) == 'foo'
}

struct Bar {
	foo Foo
}

fn g(bar Bar) {
	f({
		'name': json2.Any('foo'),
		'value': match bar.foo {
			string { bar.foo }
			int { bar.foo }
			f64 { bar.foo }
		}
	})
}

fn main() {
	g(Bar{
		foo: Foo(42)
	})
}
```

### Expected Behavior

Successful compile and runs without errors

### Current Behavior

```
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.11412134137970283209.tmp.c:8220: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/mclr/AppData/Local/Temp/v_0/test.11412134137970283209.tmp.c:18747: error: cannot convert 'struct main__Foo' to 'struct x__json2__Any'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.3 d898636

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.3 140c838.d898636
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2023-12-13 19:48:07

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2023.50-20-gd898636b
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2479,8 +2479,10 @@ fn (mut g Gen) expr_with_cast(expr ast.Expr, got_type_raw ast.Type, expected_typ
 					}
 				}
 			} else if expr is ast.SelectorExpr {
-				if _ := scope.find_struct_field(expr.expr.str(), expr.expr_type, expr.field_name) {
-					is_already_sum_type = true
+				if v := scope.find_struct_field(expr.expr.str(), expr.expr_type, expr.field_name) {
+					if v.smartcasts.len > 0 && unwrapped_expected_type == v.orig_type {
+						is_already_sum_type = true
+					}
 				}
 			}
 			if is_already_sum_type && !g.inside_return {

--- a/vlib/v/tests/match_smartcast_test.v
+++ b/vlib/v/tests/match_smartcast_test.v
@@ -97,3 +97,23 @@ fn test_match_mut_selector() {
 		}
 	}
 }
+
+// for issue 20167
+type Any_1 = f64 | int | string
+type Any_2 = int | string
+
+struct Struct {
+	field Any_2
+}
+
+fn test_branches_return_struct_field() {
+	any_2 := Struct{Any_2(42)}
+	m := {
+		'item1': Any_1('')
+		'item2': match any_2.field {
+			string { any_2.field }
+			int { any_2.field }
+		}
+	}
+	assert m['item2']! == Any_1(42)
+}

